tests/test_spectrograph.py

```diff
@@ -298,6 +355,146 @@ def test_get_status_getAnalogIn_fails(self):
+
+        spec = AvsFiberSpectrograph()
+        with self.assertRaisesRegex(AvsReturnError, "Measure"):
+            await spec.expose(duration)
+        self.patch.return_value.AVS_PrepareMeasure.assert_called_once()
+        self.patch.return_value.AVS_Measure.assert_called_once_with(self.handle, 0, 1)
+
+    async def test_expose_GetLambda_fails(self):
+        duration = 0.5  # seconds
+        self.patch.return_value.AVS_GetLambda.side_effect = None
+        self.patch.return_value.AVS_GetLambda.return_value = AvsReturnCode.ERR_INVALID_DEVICE_ID.value
+
+        spec = AvsFiberSpectrograph()
+        with self.assertRaisesRegex(AvsReturnError, "GetLambda"):
+            await spec.expose(duration)
+        self.patch.return_value.AVS_PrepareMeasure.assert_called_once()
+        self.patch.return_value.AVS_Measure.assert_called_once_with(self.handle, 0, 1)
+        self.patch.return_value.AVS_PollScan.assert_not_called()
+
+    async def test_expose_PollScan_fails(self):
+        duration = 0.5  # seconds
+        self.patch.return_value.AVS_PollScan.side_effect = None
+        self.patch.return_value.AVS_PollScan.return_value = AvsReturnCode.ERR_INVALID_DEVICE_ID.value
+
+        spec = AvsFiberSpectrograph()
+        with self.assertRaisesRegex(AvsReturnError, "PollScan"):
+            await spec.expose(duration)
+        self.patch.return_value.AVS_PrepareMeasure.assert_called_once()
+        self.patch.return_value.AVS_Measure.assert_called_once_with(self.handle, 0, 1)
+        self.patch.return_value.AVS_PollScan.assert_called_once_with(self.handle)
+
+    async def test_expose_GetScopeData_fails(self):
+        duration = 0.5  # seconds
+        self.patch.return_value.AVS_GetScopeData.side_effect = None
+        self.patch.return_value.AVS_GetScopeData.return_value = AvsReturnCode.ERR_INVALID_MEAS_DATA.value
+
+        spec = AvsFiberSpectrograph()
+        with self.assertRaisesRegex(AvsReturnError, "GetScopeData"):
+            await spec.expose(duration)
+        self.patch.return_value.AVS_PrepareMeasure.assert_called_once()
+        self.patch.return_value.AVS_Measure.assert_called_once_with(self.handle, 0, 1)
+        self.assertEqual(self.patch.return_value.AVS_PollScan.call_count, 4)
+
+    async def test_stop_exposure(self):
+        """Test that `stop_exposure` makes ends the active `expose`."""
+        duration = 5  # seconds
+        spec = AvsFiberSpectrograph()
+
+        t0 = time.perf_counter()
+        task = asyncio.create_task(spec.expose(duration))
+        await asyncio.sleep(0.1)  # give the event loop time to start
+        spec.stop_exposure()
+        await task
+        t1 = time.perf_counter()
+
+        self.assertLess(t1 - t0, 1)  # cancelling the task should make it end much sooner than the duration
+        self.patch.return_value.AVS_StopMeasure.assert_called_with(self.handle)
+        self.assertIsNone(task.result())
```

Comment:
Please add a test that cancels the exposure after the initial sleep is done but before the spectrograph has data available.